### PR TITLE
Add instructions for pre-build images with Github Actions

### PR DESCRIPTION
**Pull request by stuart23**
This addition was built for StyleSeat ticket [9203](https://astronomer.zendesk.com/agent/tickets/9203) but will definitely be used by other customers as well.
It has been tested to work both [internally](https://github.com/astronomer/stu-actions-test/runs/6279498630?check_suite_focus=true) and with the customer in [9203](https://astronomer.zendesk.com/agent/tickets/9203).

### CODE REVIEW

**Comment by stuart23** [comment]
I know we might convert the CI/CD options to tabs (I think that would look great). How do we want to group multiple Github actions methods?

**Comment by stuart23** [comment]
@jwitz @paolaperaza @kwcanuck this is ready for review.

File: astro/ci-cd.md [hunk 1]
```diff
@@ -162,6 +162,84 @@ This setup assumes the following prerequisites:
...
     ```
 
+### GitHub Actions (With Pre-Build Base Image)
+
+The process for consuming secrets during the build of images with dependancies in private repositories is described here: [Install Python Packages from a Private GitHub Repository](develop-project.md). Special considerations have to be made for setting up GitHub Actions workflow to build and deploy these images.
+
+This setup assumes the following prerequisites:
+
+- You have completed the instructions to [Install Python Packages from a Private GitHub Repository](develop-project.md)
+- The Private Key file that was used to authenticate to GitHub is still available.
```
[the remark this answers is not in the document]
Comment (stuart23) [reply astro/ci-cd.md:172]: You prefer "The Private Key file that _you_ used..." over "The Private Key file that _was_ used"?

**Review by jwitz — CHANGES_REQUESTED** [review]
Jumped in and added some tabs here @stuart23 , but before I do any inline editing I wanted to see if you could generalize the setup in light of your other PR (https://github.com/astronomer/docs/pull/597). Right now it looks based on the private GitHub repo setup, can we change this so that it works for either private package setup? Either by using comments in the YAML or by having more general variable names that are not GitHub dependent.

**Comment by jwitz** [comment]
@stuart23 Is this ready for review again? There's some changes to the CLI coming Tuesday that might change the commands included in this pipeline, so probably won't merge until then. But I can give you quick review and make sure it's ready to go

**Comment by stuart23** [comment]
Hey @jwitz the body of it is ready. I would like a second opinion on the readability - especially given that it is covering two potential use cases (private Github repo & private Pypi index) in one instruction set. The copy can also be tweaked if you want.

**Comment by jwitz** [comment]
Sweet, gonna wait for the dust to settle on the CLI front and then I'll give this another look copy-wise

File: astro/ci-cd.md [hunk 2]
```diff
@@ -162,6 +175,100 @@ This setup assumes the following prerequisites:
             astrocloud deploy ${{ secrets.PROD_ASTRONOMER_DEPLOYMENT_ID }}
     ```
 
+</TabItem>
+
+<TabItem value="custom">
+
+When building images that consume secrets or build-time arguments, special considerations have to be made for setting up GitHub Actions workflows. The images are build in two stages to allow extra options to be set during the first build. Some examples of custom images that are built in two stages can be found in [Install Python Packages from Private Sources](develop-project.md#install-python-packages-from-private-sources).
+
+This setup assumes the following prerequisites:
```
Comment (jwitz) [inline astro/ci-cd.md:184]:
```suggestion
#### Prerequisites
```

File: astro/ci-cd.md [hunk 3]
```diff
@@ -162,6 +175,100 @@ This setup assumes the following prerequisites:
...
+
+- You have completed the instructions to [Install Python Packages from Private Sources](develop-project.md#install-python-packages-from-private-sources) and can build the image locally.
+- If installing from a private GitHub Repo, you have access to the private key file used to authenticate to GitHub.
+
+OR
+
+- If installing from a private PyPi Index, you have a username and password to authenticate to the index.
```
Comment (jwitz) [inline astro/ci-cd.md:191]:
```suggestion
To complete this setup, you need:

- An Astro project with a custom Runtime image. For an example implementation of this, see [Install Python Packages from Private Sources](develop-project.md#install-python-packages-from-private-sources).
```

File: astro/ci-cd.md [hunk 4]
```diff
@@ -162,6 +175,100 @@ This setup assumes the following prerequisites:
...
+          with:
+            tags: custom-<astro-runtime-image>
+            load: true
+            file: Dockerfile.build
+            ssh: |
+              github=${{ env.SSH_AUTH_SOCK }}
+        - name: Deploy to Astro
+          run: |
+            brew install astronomer/cloud/astrocloud
+            astrocloud deploy ${{ secrets.ASTRONOMER_DEPLOYMENT_ID }}
```
Comment (jwitz) [inline astro/ci-cd.md:244]:
```suggestion
            brew install astronomer/tap/astro
            astro deploy ${{ secrets.ASTRONOMER_DEPLOYMENT_ID }}
```

File: astro/ci-cd.md [hunk 5]
```diff
@@ -162,6 +175,100 @@ This setup assumes the following prerequisites:
...
+          with:
+            tags: custom-<astro-runtime-image>
+            load: true
+            file: Dockerfile.build
+            build-args: |
+              PIP_EXTRA_INDEX_URL=${{ secrets.PIP_EXTRA_INDEX_URL }}
+        - name: Deploy to Astro
+          run: |
+            brew install astronomer/cloud/astrocloud
+            astrocloud deploy ${{ secrets.ASTRONOMER_DEPLOYMENT_ID }}
```
Comment (jwitz) [inline astro/ci-cd.md:260]:
```suggestion
            brew install astronomer/tap/astro
            astro deploy ${{ secrets.ASTRONOMER_DEPLOYMENT_ID }}
```

**Review by jwitz — CHANGES_REQUESTED** [review]
@stuart23 Sorry but I'm still confused about the scope here. Originally I thought this pipeline was going to be for _all_ customizable image use cases. Instead of generalizing, it looks like we've created a pipeline template for a very specific use case.

We should either: 
- be explicit about the fact that these templates are for users building python packages from a private reg. If this is the case, then it might even be better to document the pipelines in https://docs.astronomer.io/astro/develop-project#install-python-packages-from-private-sources to keep everything in one place
- generalize this template further so that it applies to most custom image use cases. That would involve generalizing what we put in `build-args`, `ssh`, etc. We can still mention the other private reg setup here as an example, but we'd want to focus mostly on a broader range of users. 

I'll let you decide what route to take since you are closer to the users and have a better understanding of what info they're after. Let me know if you want to jump in to the PR again and make any of these changes myself, I'm available!

**Review by jwitz — APPROVED** [review]

**Pull request merged**